dcgm: 3.2.5 -> 3.3.5, etc.

de11n opened the pull request:
## Description of changes

The build was broken so I fixed it by using static yaml-cpp (which is what DCGM's own build scripts do).

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
yaml-ccp: allow overriding static build

Commit:
dcgm: fix build by using static yaml-cpp

Commit:
dcgm: 3.2.5 -> 3.3.5

Commit:
prometheus-dcgm-exporter: 3.2.5-3.1.7 -> 3.3.5-3.4.0

Changes to go.mod are no longer needed because the package upgraded to
go 1.21.

SomeoneSerge commented on pkgs/development/libraries/yaml-cpp/default.nix:
```diff
@@ -24,7 +25,7 @@ stdenv.mkDerivation rec {
   cmakeFlags = [
     "-DYAML_CPP_BUILD_TOOLS=false"
-    "-DYAML_BUILD_SHARED_LIBS=${lib.boolToString (!stdenv.hostPlatform.isStatic)}"
+    "-DYAML_BUILD_SHARED_LIBS=${lib.boolToString (!static)}"
```
`(lib.cmakeBool "YAML_BUILD_SHARED_LIBS" (!static))`

de11n replied:
Nice. I'll change. It makes the flags less beautiful though because the no longer line up. Unless you'd prefer I convert all of them to this form?

ghthor replied:
Yeah, all those flags should use `(lib.cmakeBool )` IMO; but the others are literal strings so it's fine to leave that work for someone else

de11n replied:
I don't mind changing it myself if that's preferred.

SomeoneSerge replied:
They'd get updated eventually. One would often abstain from rewriting things just for the sake of formatting so as not to inflate the diff. In this case there is only a total of 3 flags -> no excess review burden -> it's entirely your call:)

de11n replied:
I'll leave it as-is then.

SomeoneSerge commented on pkgs/os-specific/linux/dcgm/default.nix:
```diff
@@ -123,7 +124,7 @@ in gcc11Stdenv.mkDerivation rec {
     jsoncpp-static
     libevent-nossl-static
-    yaml-cpp
+    yaml-cpp-static
```
[I don't mind, but] why force static and not conditional

de11n replied:
Do you mean for `yaml-cpp` specifically or for all dependencies? Some of them must be static for DCGM's build system to work (unless we patch it). However, DCGM's build is optimized for docker-based or RPM/Deb packaging, which doesn't have the same guarantees that Nix provides. Ideally, we'd probably not use static for any dependencies.

That said, I'm not interested in maintaining patches to DCGM's CMake build system and it's closure is already enormous, so it's not a meaningful cost. If we are able to get away with dynamic dependencies in some cases, we'll still have to override them to use `gcc11Stdenv`.

SomeoneSerge replied:
I was only looking at `yaml-cpp` since that was in the diff

de11n replied:
`yaml-cpp.override { stdenv = gcc11Stdenv; }` also fixes the build, do you prefer it? My point above is that if we do this, we might as well see if we can do it to other dependencies as well. Upstream uses static linking for everything.

SomeoneSerge replied:
I think it's best that we choose between static/dynamic linkage based on `hostPlatform.isStatic`. You can rewrite the other dependencies accordingly too if you feel up to it. It would be nice if you did, but it's not necessary for this PR

de11n replied:
I'll see what I can do.

de11n replied:
I pushed some changes to parameterize static-ness on DCGM. It's a bit silly at the moment since I've confirmed that DCGM's build only allows yaml-cpp to *actually* be linked dynamically. So setting `static = false` really only affects `yaml-cpp`, but the machinery is in place that allows us to improve it over time with fewer changes.

Commit:
libevent: add static build option

Commit:
dcgm: add static build option

Currently only one dependency (yaml-cpp) actually supports being either static or dynamic because DCGM's build system hard-codes the use of static artifacts for other dependencies. We note that in the comments as future work.

SomeoneSerge commented on pkgs/os-specific/linux/dcgm/default.nix:
```diff
@@ -116,14 +108,18 @@ in gcc11Stdenv.mkDerivation rec {
+
+    # Dependencies that can be either static or dynamic.
+    (fmt_9.override { enableShared = !static; }) # DCGM's build uses the static outputs regardless of enableShared
+    (yaml-cpp.override { inherit static; stdenv = gcc11Stdenv; })
```
Why do we have to override `stdenv`? If we use `gcc11Stdenv` for `yaml-cpp` but not for the other packages it'll be linked to a different `libc` (when linked dynamically)

SomeoneSerge commented on pkgs/os-specific/linux/dcgm/default.nix:
```diff
@@ -85,13 +77,13 @@ let
 # C.f. https://github.com/NVIDIA/DCGM/blob/7e1012302679e4bb7496483b32dcffb56e528c92/dcgmbuild/build.sh#L22
 in gcc11Stdenv.mkDerivation rec {
```
Ok, I suspect this means that you have to override all other (dynamically linked) dependencies to use the same stdenv, just like you did for `yaml-cpp`. Otherwise they might link whatever libstdc++ is default in nixpkgs (gcc13). If DCGM links a libstdc++ too (I guess it does, otherwise why override yaml-cpp), it'll load the one coming with gcc11 which won't have the symbols needed by the dependencies (e.g. by yaml-cpp)...

I think a better solution is to make DGCM link the newer libstdc++. This is what we do for cudaPackages: https://github.com/NixOS/nixpkgs/blob/bfaf0a1feb46a3ef8b17a195d52ee8d4c029e28a/pkgs/development/cuda-modules/backend-stdenv.nix#L14 (disclaimer: `useLibsFrom` might get removed in the future; it's ok, we'll figure out a migration plan)

This way you won't have to override `yaml-cpp` or any other packages stdenv-wise

SomeoneSerge replied:
This translates into `useLibsFrom stdenv gcc11Stdenv`

SomeoneSerge commented on pkgs/os-specific/linux/dcgm/default.nix:
```diff
@@ -116,14 +108,18 @@ in gcc11Stdenv.mkDerivation rec {
+    (yaml-cpp.override { inherit static; stdenv = gcc11Stdenv; })
+
+    # TODO: Dependencies that DCGM's CMake hard-codes to be static-only.
+    (jsoncpp.override { enableStatic = true; })
+    (libevent.override { sslSupport = false; static = true; })
```
Ok there's actually one more concern I didn't raise before in relation to the static/dynamic packages and overrides: I think it's desirable to avoid ad hoc overrides (e.g. overrides in let-in expressions or as here) and directly consume the top-level attributes instead. This way it's easier to override dependencies and to control them from overlays. For instance, instead of an ad hoc `libevent-nossl-static` in a `let-in` we could have dedicated an attribute with the same name in `pkgs`.

I'd also check if the build works with all overrides removed. I'd test `pkgs.dcgm` and `pkgsStatic.dcgm`.

By avoiding all this overspecification in `dcgm/default.nix` we keep the package more configurable/overridable. That said, I've no idea what the original motivation behind all of these flags was, so you probably know better

de11n replied:
DCGM's CMake isn't nearly flexible enough to support that kind of arbitrary overriding. If we wanted to support that, we'd have a *lot* more work to do (namely, patching DCGM's build system in non-trivial ways). I do not have the bandwidth to find or maintain such a patchset.

SomeoneSerge commented on pkgs/os-specific/linux/dcgm/default.nix:
```diff
@@ -15,18 +15,10 @@
-  # Flags copied from DCGM's libevent build script
-  libevent-nossl = libevent.override { sslSupport = false; };
-  libevent-nossl-static = libevent-nossl.overrideAttrs (super: {
-    CFLAGS = "-Wno-cast-function-type -Wno-implicit-fallthrough -fPIC";
-    CXXFLAGS = "-Wno-cast-function-type -Wno-implicit-fallthrough -fPIC";
-    configureFlags = super.configureFlags ++ [ "--disable-shared" "--with-pic" ];
```
These flags not needed anymore?

SomeoneSerge commented on pkgs/development/libraries/yaml-cpp/default.nix:
@de11n would you like to move the hostPlatform.isStatic-related changes into a separate PR so we can merge the update sooner?

SomeoneSerge commented on pkgs/development/libraries/yaml-cpp/default.nix:
It's been a while, let's recap:

- AFAIU, removing the manual overrides stumbles on upstream failing to use `find_package` instead of hard-coding and vendoring stuff
- There's the unsolved `gcc11Stdenv` business, which is orthogonal to the update

Should we merge?

de11n replied:
Thanks for the ping. I haven't had time to look into this recently. Correct on the first point. The second point I plan to look into eventually once I have spare cycles but I do think it could be a separate PR.

SomeoneSerge commented:
@ofborg eval

name-snrl commented:
@de11n @SomeoneSerge Hi guys, I tried to install `dcgm` but build fails on any commit I choose. Commits (commit of current system and latest 24.05) from 24.05 crash with yaml-cpp linking error. Then I tried to build from `f03439c94e7b1eefe69ee376ae26213708e098a2` or latest `nixos-unstable`:

```
-- Found CUDA 10. CUDA10_INCLUDE_DIR=/nix/store/7j1g4ihvgissb8kng2wyinxrq5gykybx-cuda-combined-10/targets/x86_64-linux/include
/nix/store/5nyjv3c38qfv6n3hs6wa4fawz9vz63qv-cudatoolkit-10.2.89/lib/libcublas_static.a
/nix/store/84wqcnbdhfz4cq7i6mqh6f6lb81rrb9g-cudatoolkit-10.2.89-lib/lib/libcudart_static.a
/nix/store/5nyjv3c38qfv6n3hs6wa4fawz9vz63qv-cudatoolkit-10.2.89/lib/libculibos.a
/nix/store/5nyjv3c38qfv6n3hs6wa4fawz9vz63qv-cudatoolkit-10.2.89/lib/libcublasLt_static.a
/nix/store/5nyjv3c38qfv6n3hs6wa4fawz9vz63qv-cudatoolkit-10.2.89/lib/stubs/libcuda.so
CUDA11_INCLUDE_DIR-NOTFOUND
libcublas11-NOTFOUND
libcudart11-NOTFOUND
libculibos11-NOTFOUND
libcublaslt11-NOTFOUND
libcuda11-NOTFOUND
CMake Error at cmake/FindCuda.cmake:92 (message):
  Could NOT find Cuda 11
Call Stack (most recent call first):
  cmake/FindCuda.cmake:112 (load_cuda)
  cublas_proxy/CMakeLists.txt:14 (find_package)


-- Configuring incomplete, errors occurred!
```

I also tried to add `.override { static = true; }`, but that doesn't change anything. Am I missing something?
